### PR TITLE
[serve] Incorporate revised dag_to_dot into Gradio Visualization

### PR DESCRIPTION
**Pull request by zcin**
Signed-off-by: Cindy Zhang <cindyzyx9@gmail.com>

### Revising `_dag_to_dot`
Currently, if `vis_utils._dag_to_dot` is called on the root `DAGNode` of a DAG that has been transformed into `DeploymentExecutorNode`s through `deployment_graph_build.build`, then it will give a graph like the following:
Screen Shot 2022-08-17 at 1 34 59 PM

... while it should be like this:
image

This is because once a DAG is transformed to a serve executor DAG, there are multiple instances of each `DAGNode` with the same `uuid`. 

The function `dag_node.apply_recursive` is suited for traversing a DAG regardless of what transformation stage it is in. It makes sense to re-use `apply_recursive` to do a dfs-traversal through the graph, instead of the current, custom dfs-traversal used in `_dag_to_dot`.

#### Changes made to `InputNode`

A separate, but related issue, is that when users build their graph like so:
```python
... = f.bind(user_input[0])
... = g.bind(user_input[0])
```
... then two separate `InputAttributeNode`s are created, and correspondingly it appears as two separate nodes in the graph generated by `_dag_to_dot`. This PR also ensures the above code will only generate one `InputAttributeNode`.

### Including in Gradio Visualization
Including a graphical illustration of the graph with node dependencies complements the current Gradio visualization.

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/dag/input_node.py [hunk 1]
```diff
@@ -70,6 +70,7 @@ def __init__(self, *args, _other_args_to_resolve=None, **kwargs):
         if len(args) != 0 or len(kwargs) != 0:
             raise ValueError("InputNode should not take any args or kwargs.")
 
+        self.input_nodes = {}
```
Comment (shrekris-anyscale) [inline python/ray/dag/input_node.py:73]: Is there another variable name we can use for this? It's a bit strange to have the `InputNode` object store a reference to what seems like multiple `input_nodes`.

File: python/ray/serve/experimental/gradio_visualize_graph.py [hunk 2]
```diff
@@ -232,6 +234,20 @@ def depths_fn(node):
         }
 
         with gr.Blocks() as demo:
+            from PIL import Image
+
+            try:
+                graph = _dag_to_dot(self.dag)
+                gr.Image(
+                    label="Direct Acyclic Graph",
```
Comment (shrekris-anyscale) [inline python/ray/serve/experimental/gradio_visualize_graph.py:242]:
```suggestion
                    label="Ray Serve Deployment Graph",
```

File: python/ray/serve/experimental/gradio_visualize_graph.py [hunk 3]
```diff
@@ -232,6 +234,20 @@ def depths_fn(node):
...
+                    value=Image.open(BytesIO(graph.create(graph.prog, format="png"))),
+                )
+            except ModuleNotFoundError:
+                logger.warning(
+                    "Module `pydot` is not installed. Unable to show pydot "
```
Comment (shrekris-anyscale) [inline python/ray/serve/experimental/gradio_visualize_graph.py:247]: Do users need `graphviz` as well? Can we also include the `pip install` command needed to get the packages in this error message?

**Review by shrekris-anyscale — COMMENTED** [review]
Great work! I left some comments.

File: python/ray/dag/input_node.py (hunk 1 above)
Comment (zcin) [reply python/ray/dag/input_node.py:73]: Renamed to `input_attribute_nodes`. Lmk if there are other suggestions!

File: python/ray/serve/experimental/gradio_visualize_graph.py (hunk 3 above)
Comment (zcin) [reply python/ray/serve/experimental/gradio_visualize_graph.py]: Included both!

**Review by shrekris-anyscale — APPROVED** [review]
Nice work, this change looks good to me!

File: python/ray/serve/experimental/gradio_visualize_graph.py [hunk 4]
```diff
@@ -232,6 +234,22 @@ def depths_fn(node):
...
+                    label="Ray Serve Deployment Graph",
+                    value=Image.open(BytesIO(graph.create(graph.prog, format="png"))),
+                )
+            except ImportError:
+                logger.warning(
+                    "Python module `pydot` and package `graphviz` is needed to show "
+                    "illustration of graph. Install pydot with `pip install pydot` "
+                    "and graphviz with either `brew install pydot` or `sudo apt "
+                    "install graphviz`"
+                )
```
Comment (simon-mo) [inline python/ray/serve/experimental/gradio_visualize_graph.py:251]: can this be a gr.Label to surface to ewb ui directly?

**Review by simon-mo — APPROVED** [review]

File: python/ray/serve/experimental/gradio_visualize_graph.py (hunk 4 above)
Comment (zcin) [reply python/ray/serve/experimental/gradio_visualize_graph.py]: Changed to a gr.Markdown that shows the warning

**Comment by simon-mo** [comment]
Merge conflict!

**Pull request merged**